File: PAT_Basic_Level_by_perl/1067 试密码/1067 试密码.pl
use strict;
use warnings;

# =====================================================================
# 题目：1067 试密码
# 难度：简单
# 知识点：字符串处理, 循环
# =====================================================================
#
# 【题目描述】
#   模拟登录：给定正确密码和允许尝试次数 N。
#   输入错误密码输出"Wrong password: 密码"；
#   正确则输出"Welcome in"并结束；错误达 N 次输出"Account locked"并结束；
#   读到单个 # 时结束输入。
#
# 【实现原理】
#   1. 读取正确密码与尝试次数 N。
#   2. 逐行读取尝试，遇到 # 结束。
#   3. 密码正确输出 Welcome in 并退出；错误计数达到 N 则锁定。
#
# 【输入样例】
#   Correct%pw 3
#   correct%pw
#   Correct@PW
#   whatisthepassword!
#   Correct%pw
#   #
# 【输出样例】
#   Wrong password: correct%pw
#   Wrong password: Correct@PW
#   Wrong password: whatisthepassword!
#   Account locked
# =====================================================================

my $line = <STDIN>;
chomp $line;
my ($pw, $n) = split /\s+/, $line;
$n = $n + 0;

my $wrong = 0;
while (my $l = <STDIN>) {
    chomp $l;
    last if $l eq '#';                 # 结束标记
    if ($l eq $pw) {
        print "Welcome in\n";
        last;
    }
    print "Wrong password: $l\n";
    $wrong++;
    if ($wrong >= $n) {                # 错误次数达上限则锁定
        print "Account locked\n";
        last;
    }
}

File: PAT_Basic_Level_by_ruby/1067 试密码/1067 试密码.rb
# 1067 试密码
#
# 题目描述：
#   当你试图登录某个系统却忘了密码时，系统一般只会允许你尝试有限多次，当超出允许次数时，账号就会被锁死。本题就请你实现这个小功能。
#
# 实现原理：
#   逐次尝试输入的密码；与正确密码比较判断是否正确；处理尝试次数限制和锁定情况
#

if __FILE__ == $PROGRAM_NAME
  line = gets(chomp: true)
  parts = line.split
  attempts_allowed = parts.pop.to_i
  password = parts.join(' ')

  attempts_allowed.times do
    attempt = gets(chomp: true)
    if attempt == '#'
      exit
    elsif attempt == password
      puts 'Welcome in'
      exit
    else
      puts "Wrong password: #{attempt}"
    end
  end
  puts 'Account locked'
end
